applications/ConstitutiveLawsApplication/custom_constitutive/traction_separation_law.cpp

```diff
@@ -0,0 +1,751 @@
+
+                mDelaminationDamageModeOne[i+1] = DelaminationDamageModeOne[i+1];
+                mThresholdModeOne[i] = equivalent_stress_mode_one;
+
+            }
+
+            const double F_mode_two = equivalent_stress_mode_two - ThresholdModeTwo[i];
+            if (F_mode_two > tolerance) {
+                
+                // const double AParameter_mode_two = -std::pow(T0s, 2) / (2.0 * Gi * GIIc / characteristic_length); // Linear
+                const double AParameter_mode_two = 1.00 / (GIIc * Gi / (characteristic_length * std::pow(T0s, 2)) - 0.5); // Exponential
+        
+                KRATOS_ERROR_IF(AParameter_mode_two < 0.0) << "AParameter_mode_two is negative." << std::endl;
+
+                // DelaminationDamageModeTwo[i+1] = (1.0 - T0s / equivalent_stress_mode_two) / (1.0 + AParameter_mode_two); // Linear
+                DelaminationDamageModeTwo[i+1] = 1.0 - (T0s / equivalent_stress_mode_two) * std::exp(AParameter_mode_two *
+                    (1.0 - equivalent_stress_mode_two / T0s)); // Exponential
+
+                DelaminationDamageModeTwo[i+1] = (DelaminationDamageModeTwo[i+1] >= 0.99999) ? 0.99999 : DelaminationDamageModeTwo[i+1];
+                DelaminationDamageModeTwo[i+1] = (DelaminationDamageModeTwo[i+1] < 0.0) ? 0.0 : DelaminationDamageModeTwo[i+1];
+
+                mDelaminationDamageModeTwo[i+1] = DelaminationDamageModeTwo[i+1];
+                mThresholdModeTwo[i] = equivalent_stress_mode_two;
+            }
+
+            // End damage calculation
+        }
+
+        for(IndexType i=0; i < p_constitutive_law_vector.size(); ++i) {
+            double layer_damage_variable_mode_one = 0;
+            double layer_damage_variable_mode_two = 0;
+
+            if (DelaminationDamageModeOne[i+1] > DelaminationDamageModeOne[i]) {
+                layer_damage_variable_mode_one = DelaminationDamageModeOne[i+1];
+            } else {
+                layer_damage_variable_mode_one = DelaminationDamageModeOne[i];
+            }
+
+            if (DelaminationDamageModeTwo[i+1] > DelaminationDamageModeTwo[i]) {
+                layer_damage_variable_mode_two = DelaminationDamageModeTwo[i+1];
+            } else {
+                layer_damage_variable_mode_two = DelaminationDamageModeTwo[i];
+            }
+
+            layer_stress[i][2] *= ((1.0-layer_damage_variable_mode_one));
+            layer_stress[i][4] *= ((1.0-layer_damage_variable_mode_one) * (1.0-layer_damage_variable_mode_two));
+            layer_stress[i][5] *= ((1.0-layer_damage_variable_mode_one) * (1.0-layer_damage_variable_mode_two));
+            layer_stress[i][0] *= 1.0;
+            layer_stress[i][1] *= 1.0;
+            layer_stress[i][3] *= 1.0;
+        }
+        
+
+        for(IndexType i=0; i < p_constitutive_law_vector.size(); ++i) {
+            const double factor = combination_factors[i];
+            delamination_damage_affected_stress_vector += factor * layer_stress[i];
```

Comment:
I think that in the `FinalizeMaterialResponse` is not necessary to compute the final stress... (not used)

Reply:
Done